### PR TITLE
Don't output implicit defaults from "config list"

### PR DESCRIPTION
**Pull request by dbarnett**
Accept an --include-defaults arg to enable including those.

Listing a nonexistent name is no longer an error, but does output a warning to stderr when no config entries match to explain why there's no other output.

This is another "config list" improvement related to #531.

# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [x] I have added tests to cover my changes

### CODE REVIEW

File: src/cli_util.rs [hunk 1]
```diff
@@ -276,6 +277,7 @@ impl CommandHelper {
         string_args: Vec<String>,
         global_args: GlobalArgs,
         settings: UserSettings,
+        layered_configs: LayeredConfigs,
```
Comment (dbarnett) [inline src/cli_util.rs:280]: Clippy doesn't like this 8th argument. Could you suggest how to restructure those (or where to plumb through these layered_configs if not here).

It's partially redundant with UserSettings, but consolidating those got a little messy in the calling code.

Comment (martinvonz) [reply src/cli_util.rs:280]: Feel free to tell Clippy to ignore it for now.

**Comment by dbarnett** [comment]
Please have a good look through for readability & mechanics on this one. Suggestions very welcome.

File: src/commands.rs [hunk 2]
```diff
@@ -1239,25 +1242,43 @@ fn cmd_config(
...
-                None => settings.config().collect()?.into(),
-            };
-            write_config_entry(ui, name.as_deref().unwrap_or(""), raw_values)?;
+                if let Some(name) = name {
+                    // Collect any match into values.
+                    match config.get::<config::Value>(name) {
+                        Ok(val) => values.push(val),
+                        Err(config::ConfigError::NotFound { .. }) => {}
+                        Err(e) => return Err(e.into()),
+                    }
```
Comment (dbarnett) [inline src/commands.rs:1263]: Felt like there might be a helper to simplify this part better, to convert the `Result<Value>` to a `Result<Option<Value>>` while swallowing the `NotFound` special case or something like that.

File: src/commands.rs [hunk 3]
```diff
@@ -179,8 +179,11 @@ enum ConfigSubcommand {
         /// An optional name of a specific config option to look up.
         #[arg(value_parser=NonEmptyStringValueParser::new())]
         name: Option<String>,
+        /// Whether to explicitly include built-in default values in the list.
+        #[arg(long)]
+        include_defaults: bool,
```
Comment (dbarnett) [inline src/commands.rs:184]: This name `--include-defaults` seem reasonable?

File: src/commands.rs [hunk 4]
```diff
@@ -1239,25 +1242,43 @@ fn cmd_config(
     command: &CommandHelper,
     subcommand: &ConfigSubcommand,
 ) -> Result<(), CommandError> {
-    ui.request_pager();
```
Comment (yuja) [inline src/commands.rs:1242]: Split this change to separate commit? It's a bug fix for `jj config edit`.

It might be also good to split cmd_config_list/_edit() function since they are quite different in implementation.

File: tests/test_config_command.rs [hunk 5]
```diff
@@ -194,13 +217,18 @@ fn test_config_layer_override_env() {
...
     ui.editor="repo"
     "###);
 
     // Environment override
     test_env.add_env_var("JJ_EDITOR", "env-override");
     let stdout = test_env.jj_cmd_success(&repo_path, &["config", "list", config_key]);
     insta::assert_snapshot!(stdout, @r###"
+    ui.editor="env-base"
+    ui.editor="user"
+    ui.editor="repo"
```
Comment (yuja) [inline tests/test_config_command.rs:231]: > Also list all values for a given config key name (ordered by precedence)

What's the main point of this change? I think printing overridden values is rather confusing.

File: src/commands.rs (hunk 2 above)
Comment (yuja) [reply src/commands.rs:1263]: Yeah, it's annoying, but I couldn't find a library function for that.

Maybe we need to add a helper function or extension trait (like `IoResultExt`.

File: tests/test_config_command.rs (hunk 5 above)
Comment (dbarnett) [reply tests/test_config_command.rs:231]: That's how `git config -l` works. Since the command is called "list" it makes sense to me that it would report multiple (and not consider it an "error" if there are zero). Also if we're including options like `--show-origin` it's helpful to include that original context.

We did talk about whether it makes sense to have a "get" vs "list" command a while back, and I _would_ expect a "get" command to report only the final value. I'd also envisioned colorizing this output eventually and shading overridden values a muted gray, plus maybe an option for whether to list overridden values.

Comment (yuja) [reply tests/test_config_command.rs:231]: I considered it a "list" matching the given name or pattern, not a list of values including overridden ones. Maybe I'm too familiar with hg than git.

Colorizing output isn't always doable, so it might be better to comment out overridden values. It also conforms to TOML where duplicated keys are invalid.
```
# ui.editor="env-base"
# ui.editor="user"
ui.editor="repo"
```
(But I still prefer not showing overridden values by default.)

Comment (dbarnett) [reply tests/test_config_command.rs:231]: Ooh, I like that idea commenting out.

FWIW another analog is Chrome Devtools will list all CSS rules with a strikethrough for overridden ones.

Anyway let's get some input from @martinvonz and others. I could imagine putting those all behind an `--include-overridden` option, but that could also be saved as an improvement to add as follow-up.

Comment (martinvonz) [reply tests/test_config_command.rs:231]: I also like the idea of commenting them out so it's valid TOML, but also so it's clear that they're not in effect. I think we should still comment it out if the user did `jj config list --user ui.editor` in the example above.

I think we should still default to not including overridden values. The option to see them seems useful for troubleshooting, but I feel like I've almost always wanted to just see the applied config when I've used `hg config` in the past. Do you know why git includes them by default, or do you have a usecase for it?

Comment (dbarnett) [reply tests/test_config_command.rs:231]: Alright, I can just exclude them for now and work on an `--include-overridden` option later (showing them commented-out).

Still thinking nonexistent names should be treated as overall "success" but with a hint to stderr, SG?

And one other wrinkle I thought of: I'd kind of expect built-in default values to not show *unless* you'd specifically requested those names. E.g. if I didn't have any explicit `colors` values configured their defaulted values wouldn't show up in `config list` but maybe should in `config list colors`. Opinions on that?

Comment (martinvonz) [reply tests/test_config_command.rs:231]: Sounds good.

I would probably not include built-in default values without a `--include-defaults` flag, especially since it depends so much on exactly how we've implemented the defaults - some defaults are applied to the `LayeredConfigs`, and some are applied at the use site. Treating `config list colors` specially seems like it would lead to confusion.

Comment (dbarnett) [reply tests/test_config_command.rs]: Okay, the good news is I got the behavior of hiding overridden values implemented. The bad news is my code for doing it looks... just terrible. (Although that's also potentially good news for me, because I'm probably going to learn a lot getting suggestions from more experienced rustaceans about how to do it properly 😁).

Please give me some suggestions to fix this mess. I'll leave a few inline comments about parts where I got stuck...

File: src/commands.rs (hunk 2 above)
Comment (dbarnett) [reply src/commands.rs]: K, as long as there's not an obvious tidy way to do it, I'll resolve this comment for now and keep an eye out for a way to simplify it in future changes.

File: src/cli_util.rs [hunk 6]
```diff
@@ -310,6 +314,16 @@ impl CommandHelper {
         &self.settings
     }
 
+    pub fn resolved_config_values(
+        &self,
+        name: Option<&str>,
+        include_defaults: bool,
+    ) -> Result<Vec<(ConfigSource, Option<String>, config::Value)>, crate::config::ConfigError>
```
Comment (dbarnett) [inline src/cli_util.rs:321]: This return type is ugly. Maybe it should return a struct or something?

File: src/commands.rs [hunk 7]
```diff
@@ -1243,22 +1245,25 @@ fn cmd_config(
...
-                                user_error("key not found in config")
-                            }
-                            _ => e.into(),
-                        })?
+        ConfigSubcommand::List {
+            name,
+            include_defaults,
+        } => {
+            let values = command
+                .resolved_config_values(name.as_ref().map(Deref::deref), *include_defaults)?;
```
Comment (dbarnett) [inline src/commands.rs:1253]: The ref manipulations here seem awkward. Is there a cleaner way to handle the refs/types?

File: src/config.rs [hunk 8]
```diff
@@ -85,23 +99,104 @@ impl LayeredConfigs {
...
+                // Push whole config onto values.
+                values.push((source, None, config.collect()?.into()));
+            }
+        }
+
+        // Walk through configs in reverse and remove overridden values from lower-precedence configs.
+        // Note: This could almost be done by merging into one large config, except that would lose source/ordering information.
+        let mut seen_config_names = HashSet::new();
+        let mut overridden_idxs = HashSet::new();
+        for (i, (_, name, value)) in values.iter_mut().enumerate().rev() {
```
Comment (dbarnett) [inline src/config.rs:155]: This really feels like it should use `retain_mut` or similar, but I keep running into issues propagating errors out of closures. Is there a good way to deal with the fallible parts and fit it into `retain_mut`?

File: src/config.rs [hunk 9]
```diff
@@ -85,23 +99,104 @@ impl LayeredConfigs {
...
+            if let config::ValueKind::Table(_) = value.kind {
```
Comment (dbarnett) [inline src/config.rs:156]: This logic is kinda redundant with the recursive helper below, except this needs to handle a series `Value`s of any type where the helper just operates on Tables. Please suggest any obvious ways to clean up the duplication.

File: src/config.rs [hunk 10]
```diff
@@ -85,23 +99,104 @@ impl LayeredConfigs {
...
+                *value = LayeredConfigs::remove_overridden_config_names(
+                    &value.to_owned().into_table()?,
```
Comment (dbarnett) [inline src/config.rs:158]: This borrowing is weird. It doesn't need to be mutable or owned AFAIK but when I try to call `into_table` directly it says I'm moving a value I shouldn't and something about `self`.

File: src/commands.rs (hunk 7 above)
Comment (yuja) [reply src/commands.rs:1253]: You can use `.as_deref()`, but I think `resolved_config_values()` does too much things.

Maybe we can instead add `layered_configs.get_source(key: &str) -> Option<ConfigSource>` which looks up `sources()` in reverse order and returns the first match. Since `write_config_entry()` decomposes intermediate table recursively, we can get the source of the leaf value there, and omit if it was `ConfigSource::Default`.

It wouldn't be as efficient as your implementation, but the whole config entries wouldn't be big.

File: src/config.rs (hunk 10 above)
Comment (dbarnett) [reply src/config.rs]: Resolved. Figured out how to use `ref` to get the table value from the pattern match, without all the weird borrowing gyrations.

File: src/config.rs (hunk 8 above)
Comment (dbarnett) [reply src/config.rs]: Yes! Managed to clean it up to use `retain_mut` after simplifying away some impossible error cases (unpacking table value with `ref`).

Although apparently `retain_mut` is a nightly feature that fails some of the tests here, so I'll still need to tweak the approach w/ a different helper...

Comment (dbarnett) [reply src/config.rs]: Yep, got an equivalent working with just `.rev()` and `.filter_map()`.

File: src/cli_util.rs (hunk 6 above)
Comment (dbarnett) [reply src/cli_util.rs]: Encapuslated into an `AnnotatedConfigEntry` struct.

File: src/commands.rs (hunk 7 above)
Comment (dbarnett) [reply src/commands.rs]: Hmm, I agree it did too many things, but having entirely separate implementations for getting the value and the source feels icky to me, not just because of performance but because it needs to duplicate some lookup logic and there could be discrepancies with how it handles precedence.

I did move the `--include-default` handling entirely out of the helper, though, and I suspect I'll be able to simplify down the logic for overridden values a little further eventually.

Comment (yuja) [reply src/commands.rs]: > there could be discrepancies with how it handles precedence.

That's unavoidable since we rely on `Config::builder()` to merge configs to be actually used, right? We can make sure that each `(value, source)` pair in `resolved_config_values()` is consistent, but it doesn't mean the value would actually be picked by `user_settings.config().get()`.

File: src/config.rs [hunk 11]
```diff
@@ -47,6 +50,23 @@ pub struct LayeredConfigs {
...
+    Repo,
+    CommandArg,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct AnnotatedConfigEntry {
+    pub name: Option<String>,
+    pub value: config::Value,
+    pub source: ConfigSource,
+}
```
Comment (yuja) [inline src/config.rs:68]: Keeping the top-level `source` wouldn't be enough. Suppose we'll add an option to show `source`, we'll need a tree of `(value, source)`, not a tree of `value` and the top-level `source`.

```
struct Annotated { .., value, source }
enum Value {
    String(..),
    ..
    Table(HashMap<String, Annotated>),
}
```

FWIW, `config::Value` is structured this way, but its `origin: Option<String>` is private and unreliable.

**Comment by martinvonz** [comment]
I'll leave this review for you, @yuja, but let me know if you want me to review it too.

File: src/commands.rs (hunk 7 above)
Comment (dbarnett) [reply src/commands.rs]: Yes, but feels slightly more bizarre to me if discrepancies could lead to reporting the right value with the wrong source. Obviously most of this logic ultimately ought to be handled by the config crate, but for now I'd really prefer to not have such piecemeal logic for values vs sources or it's gonna get ugly fast trying to support `config list --include-overridden --show-origin`.

Anyway, I realized I should be able to simplify way down by splitting out the operations doing key name lookup. It'll take me a bit to rework the code but I think it'll look much much cleaner. Stay tuned...

File: src/config.rs (hunk 11 above)
Comment (dbarnett) [reply src/config.rs:68]: Yep right, since the final resolved config can have a single table getting keys from different sources. Fixing that part too.

**Comment by dbarnett** [comment]
Thanks for sticking with me on review here. I haven't had a lot of time to get the PR updated the last couple days but have been turning it over in my mind and know how I want to clean it up. Will bump for re-review in a day or two.

File: src/commands.rs (hunk 7 above)
Comment (yuja) [reply src/commands.rs]: > Obviously most of this logic ultimately ought to be handled by the config crate, [...]

Yep, so it would be nice if we can come up with a simple workaround for the moment. I would prioritize "right value" than "wrong source", but I'm fine with any solution that looks obviously correct.

Comment (dbarnett) [reply src/commands.rs]: K, now the code's separated out for resolving the annotated config values vs. filtering the key names. The logic is kinda complex, and hopefully most of it's temporary as the config crate adds more support for these kinds of operations, but it does alright at providing the desired behavior.

**Comment by dbarnett** [comment]
Alright, I've updated the behavior. The logic is a bit complex, but if you're not opposed to the general approach I'd love any suggestions to make my Rust more idiomatic.

File: src/config.rs [hunk 12]
```diff
@@ -85,23 +131,87 @@ impl LayeredConfigs {
...
+                ),
+                _ => AnnotatedValue::Value {
+                    value,
+                    source: source.to_owned(),
+                    is_overridden: false,
+                },
+            }
+        }
+        for (source, config) in self.sources() {
+            configs.push(annotate(config.collect()?.into(), &source));
```
Comment (yuja) [inline src/config.rs:178]: We can process subtree of the given `name`. Or did you intentionally make `name` match any subtree key? 

Maybe I said `resolved_config_values(name: Option<&str>)` does too much stuff before, but I was wrong. It makes sense to narrow the config tables to process here.

File: src/cli_util.rs [hunk 13]
```diff
@@ -1394,45 +1404,22 @@ pub fn write_commit_summary(
...
+                .sorted_by_key(|(k, _)| *k)
                 .map(|(k, v)| format!("{k}={}", serialize_config_value(v)))
                 .join(", ")
         ),
-        config::ValueKind::Array(vals) => format!(
-            "[{}]",
-            vals.into_iter().map(serialize_config_value).join(", ")
-        ),
-        config::ValueKind::String(val) => format!("{val:?}"),
+        config::ValueKind::Array(ref vals) => {
```
[the remark this answers is not in the document]
Comment (samueltardieu) [reply src/cli_util.rs:1419]: The non-reference pattern matching was added to Rust to improve ergonomics, so I would suggest using `match &value.kind` as @yuja suggests and drop the `ref`.

File: src/cli_util.rs [hunk 14]
```diff
@@ -1394,45 +1404,22 @@ pub fn write_commit_summary(
...
-    };
-    Ok(())
-}
-
 // TODO: Use a proper TOML library to serialize instead.
-fn serialize_config_value(value: config::Value) -> String {
+pub fn serialize_config_value(value: &config::Value) -> String {
     match value.kind {
-        config::ValueKind::Table(table) => format!(
+        config::ValueKind::Table(ref table) => format!(
```
[the remark this answers is not in the document]
Comment (dbarnett) [reply src/cli_util.rs:1410]: In a single config I think arrays of tables are extended just like any other array, not merged or anything.

But since table values can be overridden at any level I think we need to recurse over the nested tables and give each value a place to track its source & override state.

File: src/commands/mod.rs [hunk 15]
```diff
@@ -1023,22 +1027,61 @@ fn cmd_config(
...
-                                user_error("key not found in config")
+        ConfigSubcommand::List {
+            name,
+            include_defaults,
+        } => {
+            let configs = command.resolved_config_values()?;
+            // Key name must either equal "{name}" or match "{name}.*".
+            let name_re = name
+                .as_ref()
+                .map(|name| Regex::new(format!(r"{name}(\.|$)").as_str()).unwrap());
```
[the remark this answers is not in the document]
Comment (dbarnett) [reply src/commands/mod.rs:1038]: Good idea, I can keep the string pieces separate and join them at the end. Will update.

File: src/config.rs (hunk 12 above)
Comment (dbarnett) [reply src/config.rs:178]: Yup, I like having some separation of concerns but it's a shame to materialize the entire config structure when most of it will be ignored to do it. I'll see if I can find a happy medium.

I was imagining some kind of generator/coroutine structure to decouple different aspects of the iteration more elegantly, but my rust-fu isn't strong enough for that yet, even if it would be a reasonable approach. 😅

File: src/config.rs [hunk 16]
```diff
@@ -27,6 +30,49 @@ pub enum ConfigError {
...
+                    .iter()
+                    .sorted_by_key(|(k, _)| *k)
+                    .flat_map(move |(k, v)| {
+                        let key = prefix
+                            .to_owned()
+                            .map_or(k.to_owned(), |n| format!("{n}.{k}"));
+                        v.iter_recursive(Some(key))
+                    }),
+            ),
+        }
```
[the remark this answers is not in the document]
Comment (dbarnett) [reply src/config.rs:71]: Not sure I follow this suggestion. Is it still relevant along with other suggested changes to narrow config tables?

File: src/commands/mod.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (dbarnett) [reply src/commands/mod.rs]: Yeah, that one's a huge simplification, thanks! And it allows pointing to the original string pieces as `&str` instead of making so many copies.

**Comment by dbarnett** [comment]
Updated with some quick suggested simplifications. Still planning to simplify the overall iteration logic and hoping you could clarify the part about flattening the tree to a vector.

File: src/cli_util.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (yuja) [reply src/cli_util.rs]: Because the caller flattens nested tables to filter out `source == Default`, `serialize_config_value()` wouldn't receive a value of `ValueKind::Table`. This turned out to be wrong, but made me think if we need to test the source of the table items in an array.

Anyway, an array of tables in `~/.jjconfig.toml` appears to be overridden by the one in `.jj/repo/config.toml`, so no source tests would be needed for the inner tables.

File: src/config.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (yuja) [reply src/config.rs:71]: I meant something like this might be easier to process than recursive structure:
```
[
    (["git", "pull"], Value::String("origin"), Source::Default),
    (["git", "pull"], Value::String("upstream"), Source::User),
    (["ui", "pager"], Value::Array(["less"]), Source::User),
]
```

Default values can be excluded by `.iter().filter(|(.., source)| ...`, and entries that share the same key (or maybe key prefix?) with the latter should be overridden.

I didn't consider that thoroughly, so I might be wrong.

File: src/config.rs [hunk 17]
```diff
@@ -85,23 +106,82 @@ impl LayeredConfigs {
...
+                        // TODO: Remove sorting when config crate maintains deterministic ordering.
+                        for (k, v) in table.iter().sorted_by_key(|(k, _)| *k).rev() {
+                            let mut key_path = path.to_vec();
+                            key_path.push(k);
+                            config_stack.push((key_path, v));
+                        }
+                    }
+                    _ => {
+                        config_vals.push(AnnotatedValue {
+                            path: path.iter().map(|s| s.to_string()).collect_vec(),
```
Comment (dbarnett) [inline src/config.rs:160]: I don't love cloning to `String`s here, but had trouble figuring out lifetimes to use `&str`. Any suggestions? Or is it more trouble than it's worth?

File: src/commands/mod.rs [hunk 18]
```diff
@@ -171,8 +170,11 @@ enum ConfigSubcommand {
         /// An optional name of a specific config option to look up.
         #[arg(value_parser=NonEmptyStringValueParser::new())]
         name: Option<String>,
-        // TODO(#531): Support --show-origin using LayeredConfigs.
-        // TODO(#531): Support ConfigArgs (--user or --repo) and --all.
+        /// Whether to explicitly include built-in default values in the list.
+        #[arg(long)]
+        include_defaults: bool,
```
Comment (dbarnett) [inline src/commands/mod.rs:175]: BTW I'm thinking a little about how `--include-defaults` would combine with `--user`/`--repo`/etc... Simplest is probably to not let them be combined, have them all mutually exclusive with each other and with `--include-defaults`. I could imagine supporting a union like `config list --include-defaults --env --user`, but it feels a little awkward and I doubt anyone would ever need it.

File: src/cli_util.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (dbarnett) [reply src/cli_util.rs]: Alright, I simplified and removed the Table case where it's no longer necessary.

File: src/config.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (dbarnett) [reply src/config.rs]: SG, went with an approach along those lines and flattened out the repeated recursion.

File: src/config.rs [hunk 19]
```diff
@@ -85,23 +106,82 @@ impl LayeredConfigs {
...
+    pub fn resolved_config_values(
+        &self,
+        filter_prefix: &[&str],
+    ) -> Result<Vec<AnnotatedValue>, ConfigError> {
+        // Collect annotated values from each config.
+        let mut config_vals = vec![];
+
+        for (source, config) in self.sources() {
+            let config_table = config.collect()?.into();
+            let mut config_stack: Vec<(Vec<&str>, &config::Value)> = vec![(vec![], &config_table)];
```
Comment (dbarnett) [inline src/config.rs:143]: FYI I changed to use a stack instead of recursion here simply because I lost my battle with the borrow checker trying to make something recursive and safe that was halfway readable. 😅

**Comment by dbarnett** [comment]
Alright, I consolidated down to encapsulate the logic more nicely in `resolved_config_values`. Ready for another look.

File: src/config.rs [hunk 20]
```diff
@@ -85,23 +106,82 @@ impl LayeredConfigs {
...
+            while let Some((path, value)) = config_stack.pop() {
+                // Skip if path doesn't match filter_prefix.
+                if filter_prefix.iter().zip(path.iter()).any(|(a, b)| *a != *b) {
+                    continue;
```
Comment (yuja) [inline src/config.rs:147]: Instead, the initial `config_stack` can be narrowed to the `filter_prefix`.

```rust
filter_prefix.iter().try_fold(config_table, |value, &key| {
                value
                    .into_table()
                    .ok()
                    .and_then(|mut table| table.remove(key))
            })
```

If we could build a `config::path::Expression` safely, it should be just `config.get(path)`, but there's no public API to convert names vec to the path expression.

File: src/config.rs [hunk 21]
```diff
@@ -340,4 +420,111 @@ mod tests {
...
+            env_base: env_base_config,
+            user: None,
+            repo: Some(repo_config),
+            env_overrides: empty_config,
+            arg_overrides: None,
+        };
+        assert_eq!(
+            layered_configs.resolved_config_values(&[]).unwrap(),
+            [
+                // Note: "email" is alphabetized, before "name" from same layer.
```
Comment (yuja) [inline src/config.rs:464]: Nit: You might want to use `insta::assert_debug_snapshot!()`

File: src/config.rs (hunk 17 above)
Comment (yuja) [reply src/config.rs:160]: Since `config_table` is instantiated locally, it wouldn't be possible to return `&str` that references the `config_table`. Maybe you can eliminate some of `.to_string()` by consuming the source `ValueKind::Table()`, but that wouldn't worth the effort.

File: src/commands/mod.rs (hunk 18 above)
Comment (yuja) [reply src/commands/mod.rs:175]: I think exclusive options are good. I agree it's unlikely user would want to see `--env --repo`.

**Review by yuja — APPROVED** [review]
This version looks way simpler and nice.

File: src/config.rs [hunk 22]
```diff
@@ -85,23 +106,82 @@ impl LayeredConfigs {
...
+        for val in &mut config_vals {
+            // Mark override on any previous value with same key.
+            if let Some(AnnotatedValue {
+                ref mut is_overridden,
+                ..
+            }) = keys_to_override.insert(val.path.to_owned(), val)
+            {
+                *is_overridden = true;
+            }
+        }
```
Comment (samueltardieu) [inline src/config.rs:181]: I think I prefer the more concise (and easier to read IMO):

```rust
        for val in &mut config_vals {
            // Mark override on any previous value with same key.
            if let Some(mut value) = keys_to_override.insert(val.path.to_owned(), val) {
                value.is_overridden = true;
            }
        }
```

Comment (yuja) [reply src/config.rs:181]: And mark overridden from the top layer?

```rust
        let mut keys_found = HashSet::new();
        for val in config_vals.iter_mut().rev() {
            val.is_overridden = !keys_found.insert(&val.path);
        }
```

Comment (samueltardieu) [reply src/config.rs:181]: Even better indeed.

File: src/config.rs (hunk 17 above)
Comment (dbarnett) [reply src/config.rs:160]: The catch is I'd have to explicitly deref to use `to_owned()` since it's a `&&str` where `s.to_string()` handles all the derefs implicitly. Is there a way to de-uglyify `path.iter().map(|s| (*s).to_owned()).collect_vec()` (without making throwaway copies of the whole vector)?

Comment (yuja) [reply src/config.rs:160]: `|&s| s.to_owned()` is one option. `.into_iter()` should also work, but clippy might complain.

fwiw, I agree `.to_owned()` is preferred so long as the type can be inferred.

File: src/config.rs [hunk 23]
```diff
@@ -27,6 +30,24 @@ pub enum ConfigError {
...
+    User,
+    Repo,
+    CommandArg,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct AnnotatedValue {
+    pub path: Vec<String>,
+    pub value: config::Value,
+    pub source: ConfigSource,
```
[the remark this answers is not in the document]
Comment (dbarnett) [reply src/config.rs:47]: Yep, I like it! Might do that soon.

File: src/config.rs [hunk 24]
```diff
@@ -27,6 +30,24 @@ pub enum ConfigError {
     AmbiguousSource(PathBuf, PathBuf),
 }
 
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ConfigSource {
+    Default,
+    Env,
+    // TODO: Track explicit file paths when user config is a dir.
```
[the remark this answers is not in the document]
Comment (dbarnett) [reply src/config.rs]: Yep that was my intention. Tweaked the wording.

File: src/config.rs (hunk 21 above)
Comment (dbarnett) [reply src/config.rs]: I did indeed! Thanks!

File: src/config.rs (hunk 20 above)
Comment (yuja) [reply src/config.rs:147]: Nope, to initialize the `config_stack = [prefixed_root_entry]` set.

File: src/commands/mod.rs [hunk 25]
```diff
@@ -1024,22 +1026,42 @@ fn cmd_config(
...
-                            _ => e.into(),
-                        })?
+        ConfigSubcommand::List {
+            name,
+            include_defaults,
+        } => {
+            let name_path = name
+                .as_ref()
+                .map_or(vec![], |name| name.split('.').collect_vec());
+            let values = command.resolved_config_values(&name_path)?;
```
Comment (samueltardieu) [inline src/commands/mod.rs:1036]: Maybe just make `values` mutable and add:

```rust
values.sort_unstable_by(|a, b| a.path.cmp(&b.path));
```

so that values are printed in order.

Comment (dbarnett) [reply src/commands/mod.rs:1036]: Oh, but at this level sorting the entire list would be a misfeature. It's sorted by (layer, key) and I'd like to not disturb that existing ordering by logical layer.

File: src/config.rs (hunk 20 above)
Comment (dbarnett) [reply src/config.rs:147]: But then that's basically copy-pasting the table logic from below, isn't it? If so, I feel like the miniscule performance gain wouldn't be worth the extra code bloat of unrolling the first iteration.

Comment (dbarnett) [reply src/config.rs]: Ohhhh, nevermind, you're saying instead of iterating the whole config and filtering I might as well get the value directly from the config, since at the config level it can't have multiple values for the same key anyway. Sounds good, done.

**Comment by dbarnett** [comment]
Dangit, looks like I broke the tests again. Will figure out what happened in a bit.

**Comment by dbarnett** [comment]
> Dangit, looks like I broke the tests again.

Oh, just needed to handle `ConfigError::NotFound` now that I'm calling `get` directly. Slightly less pretty now, but should be good now.

File: src/commands/mod.rs (hunk 25 above)
Comment (samueltardieu) [reply src/commands/mod.rs:1036]: I understand that, but why is it useful? For example, here is my `jj config list` for the current repository:

```
ui.editor="vim"
ui.graph.style="curved"
user.email="sam@rfc1149.net"
user.name="Samuel Tardieu"
git.fetch="upstream"
```

The `git.fetch` feels oddly (to me) out of order at the end. It is there because it comes from the repository configuration instead of the user one, but nothing indicates that.

But this can always be discussed and fixed if needed later if you want to get those changes in.

Comment (martinvonz) [reply src/commands/mod.rs:1036]: FWIW, Mercurial sorts them just by key name, and Git sorts them by source and then by key name. I don't know which I would prefer.

Comment (dbarnett) [reply src/commands/mod.rs:1036]: Right, every tool seems to do it differently, not sure there's any clearly "superior" ordering as long as it's deterministic.

Git orders entirely in order of definition/precedence, which makes the output more intuitive for options like its `--show-origin` and when it's listing overridden values. Mercurial seems to alphabetize it completely at the *top* level, but then keep nested keys in definition order.

Granted, if your config is all in one file except for a single entry having it 4/5 alphabetized looks a little mysterious. Fully alphabetizing would keep some related options close together but also splits up related options like "colors" and "ui". It would also be possible to maintain a logical ordering for known config keys that could sort of correspond to the order they're introduced in https://github.com/martinvonz/jj/blob/main/docs/config.md, but that's complicated and I felt like alphabetizing wasn't actually an improvement over retaining the precedence ordering.

Comment (dbarnett) [reply src/commands/mod.rs:1041]: K, seems like I can call the ordering question resolved and revisit later if needed, but LMK any other concerns. Will merge a bit later today.
